bookkeeper-server/src/test/java/org/apache/bookkeeper/client/DeferredSyncTest.java

```diff
@@ -50,9 +50,9 @@ public void testAddEntryLastAddConfirmedDoesNotAdvance() throws Exception {
                 .withWriteFlags(WriteFlag.DEFERRED_SYNC)
                 .execute())) {
             for (int i = 0; i < NUM_ENTRIES - 1; i++) {
-                result(wh.appendAsync(DATA));
+                result(wh.appendAsync(DATA.retainedDuplicate()));
```

Comment:
In theory this code is safe and it is an "application" that is calling this API as expected.

the fact the you have changed this test is an hint about an error which you are introducing.

Reply:
@dlg99  regarding this test I think that the application in this case is already retaining *once* the DATA ByteBuf, so the test is not to be changed.

same for DL

Reply:
@eolivelli assuming we agree that "the party who accesses a reference-counted object lastly is responsible for the destruction of the reference-counted object":

static final ByteBuf DATA = Unpooled.wrappedBuffer("foobar".getBytes());
created DATA with refCnt == 1

1st call to
result(wh.appendAsync(DATA));
will work ok, take ownership of 1 refCnt of DATA and will reduce DATA's refCnt to 0 when no longer needs it.
2nd call to 
result(wh.appendAsync(DATA));
will fail.

It is up to the calling method to decide whether object has to be retained and reused in caller's context hence 
result(wh.appendAsync(DATA.retainedDuplicate()));
Could as well do
DATA.retain();
result(wh.appendAsync(DATA));

Normally at the end of the test we should release DATA. As far as I see, netty recommends for unit tests the following:
static final ByteBuf DATA = ReferenceCountUtil. releaseLater( Unpooled.wrappedBuffer("foobar".getBytes()) );
to release it automatically.

Now let's imagine that appendAsync does not take ownership but calls retain() internally. 
Now in simple prod case when data is not used after write need to do something like that:

ByteBuf data = PooledAllocator.DEFAULT.buffer(size);
...
try {
  wh.appendAsync(data);
} finally {
  data.release();
}

Reply:
I think this is an unexpected behaviour (Maybe I need a coffee and I am missing something)

```static final ByteBuf DATA = Unpooled.wrappedBuffer("foobar".getBytes());
created DATA with refCnt == 1
1st call to
result(wh.appendAsync(DATA));
will work ok, take ownership of 1 refCnt of DATA and will reduce DATA's refCnt to 0 when no longer needs it.
```


This is not good, I (application) have a reference to DATA, I am using it, refcount = 1.
I pass it to BK, and I am continuing to hold a reference to it.
If BK uses it, it must call 'retain' any time BK wants (refcount > 1), but the net effect MUST be that when BK finished its own work the refcount is back to 1.
When my application does not need that ByteBuf anymore, it will call release() and refcount will be 0 and then Netty will free or recycle the buffer.

Is is not this the expected behaviour, that is 'net effect of BK = 0" ?

@reddycharan already brought up this discussion about the 'net effect'

Reply:
That's different from what app had before when it used i.e. byte arrays for data.
But this is expected with netty ref-counted ByteBuffs.
I'll refer you to http://netty.io/wiki/reference-counted-objects.html#who-destroys-it "Who destroys it?" chapter again.

Reply:
@dlg99 I understand your point.

Following that guide the client which is performing addEntry somehow says 'now this bytebuf is one of yours'

Reply:
@dlg99 

> "the party who accesses a reference-counted object lastly is responsible for the destruction of the reference-counted object"

I think the statement is a bit misleading. The destruction happens when the ref-count reaches 0. Each party will not know exactly when that is the case, they will just call `release()` and move on. 

I think a better way to describe the pattern is in term of "temporary ownership", or "ensuring the buffer is valid until I'm done". 

This way: 
 * If I'm passed a buffer, I assume it's a valid one
 * If I use the buffer immediately (in same thread), there's no need to retain/release
 * If I want to store a ref to the buffer and use it later (eg: pass to other thread, or keep it around in case I need to retransmit), then I need to retain/release. As a rule of thumb, I would say that this should be hidden in the API, because the caller shouldn't be forced to know what I'm going to do internally with that given buffer.

Reply:
@merlimat 

> If I'm passed a buffer, I assume it's a valid one

ok

> If I use the buffer immediately (in same thread), there's no need to retain/release

There is no need to retain, but a need to release() when you don't need it anymore. 

> If I want to store a ref to the buffer and use it later (eg: pass to other thread, or keep it around in case I need to retransmit), then I need to retain/release. 

then you release it later. when you are done with it.
when you pass it to another thread that will release it independently, do call retain() or pass retainedDuplicate() to that thread/runnable.

> As a rule of thumb, I would say that this should be hidden in the API, because the caller shouldn't be forced to know what I'm going to do internally with that given buffer

It won't be hidden as now one who passed that buffer has to wrap all calls to addAsync into try/finally and release that buffer, unlike with the rest of netty's API that uses ByteBufs.